### PR TITLE
WIP - Refactored code for common module

### PR DESCRIPTION
**Pull request by shtripat**
tendrl-spec: refactor_common_code_and_move.adoc
tendrl-bug-id:Tendrl/common#62
tendrl-bug-id:Tendrl/common#60
tendrl-bug-id:Tendrl/common#82
tendrl-bug-id:Tendrl/common#83
tendrl-bug-id:Tendrl/common#84

Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by shtripat** [comment]
@nthomas-redhat @r0h4n @brainfunked @anmolbabu @nnDarshan @GowthamShanmugam please review

**Comment by shtripat** [comment]
@nthomas-redhat @r0h4n @brainfunked @anmolbabu @nnDarshan @GowthamShanmugam please review

File: tendrl/common/atoms/base_atom.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import abc
+
+
+class BaseAtom(object):
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def run(self, parameters):
+        raise NotImplementedError(
+            'define the function run to use this class'
```
Comment (r0h4n) [inline tendrl/common/atoms/base_atom.py]: Inherit "NotImplementedError" and create "AtomNotImplementedError" (common/atoms/exceptions.py) and provide msg "run function not implemented"

File: tendrl/common/flows/base_flow.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import abc
+
+
+class BaseFlow(object):
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def run(self):
+        raise NotImplementedError(
```
Comment (r0h4n) [inline tendrl/common/flows/base_flow.py]: Inherit "NotImplementedError" and create "FlowNotImplementedError" (common/flows/exceptions.py) and provide msg "run function not implemented"

File: tendrl/common/atoms/base_atom.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import abc
+
+
+class BaseAtom(object):
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
```
Comment (r0h4n) [inline tendrl/common/atoms/base_atom.py]: BaseAtom class should contain below attributes as per Tendrl schema 0.3 (https://raw.githubusercontent.com/Tendrl/node_agent/master/tendrl/node_agent/manager/tendrl_definitions_node_agent.py)
- name (str)
- enabled (bool)
- inputs (set)
- help (str) (requires change to current definitions, change current "name" to "help")
- outputs (set)
- uuid (str)

File: tendrl/common/config.py [hunk 4]
```diff
@@ -9,22 +9,16 @@ class ConfigNotFound(Exception):
...
-            self.path = os.environ[CONFIG_PATH_VAR]
-        except KeyError:
-            self.path = DEFAULT_CONFIG_PATH
+        self.path = cfg_file_path
 
         if not os.path.exists(self.path):
-            err = ConfigNotFound("Bridge_Common Configuration not found at "
-                                 "%s" % self.path)
+            err = ConfigNotFound(
+                "Configuration not found at %s" % self.path
```
Comment (r0h4n) [inline tendrl/common/config.py]: Provide component name for whom configuration was not found in this msg

File: tendrl/common/flows/base_flow.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import abc
```
Comment (r0h4n) [inline tendrl/common/flows/base_flow.py]: BaseFlow class should contain below attributes as per Tendrl schema 0.3 (https://raw.githubusercontent.com/Tendrl/node_agent/master/tendrl/node_agent/manager/tendrl_definitions_node_agent.py)

- name (str)
- atoms (ordered list)
- help (str) (requires change to current definitions, change current "description" to "help")
- enabled (bool)
- inputs (map {mandatory: (set), optional: (set)
- pre_run (ordered list)
- post_run (ordered list)
- type (str)
- uuid (str)
- parameters (dict)

File: tendrl/common/flows/flow.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import etcd
+import logging
+
+from tendrl.common.flows.base_flow import BaseFlow
+from tendrl.common.flows import utils
+
+LOG = logging.getLogger(__name__)
+
+
+class Flow(BaseFlow):
```
Comment (r0h4n) [inline tendrl/common/flows/flow.py]: remove this class, get everything in BaseFlow

File: tendrl/common/flows/flow_execution_exception.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+class FlowExecutionFailedError(Exception):
```
Comment (r0h4n) [inline tendrl/common/flows/flow_execution_exception.py]: rename this from "flow_execution_exception" to "exceptions.py"

File: tendrl/common/manager/manager.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+import abc
```
Comment (r0h4n) [inline tendrl/common/manager/manager.py]: scrap all the current gevent.greenlet.Greenlet threads and rename them to something below
- state sync (thread which syncs state from the system under management, eg: gluster cluster, ceph etc)
- job sync (thread to process jobs from queue)

File: tendrl/common/manager/rpc.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+import yaml
+
+from tendrl.common.definitions.validator import DefinitionsSchemaValidator
+from tendrl.common.flows.flow_execution_exception import \
+    FlowExecutionFailedError
+
+LOG = logging.getLogger(__name__)
+
+
+class EtcdRPC(object):
```
Comment (r0h4n) [inline tendrl/common/manager/rpc.py]: Rename this to job sync (thread to process jobs from queue)

File: tendrl/common/persistence/persister.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    def __init__(self, fn, args, kwargs):
+        self.fn = fn
+        self.args = args
+        self.kwargs = kwargs
+
+    def call_it(self):
+        self.fn(*self.args, **self.kwargs)
+
+
+class Persister(gevent.greenlet.Greenlet):
```
Comment (r0h4n) [inline tendrl/common/persistence/persister.py]: have a persistor base class (one for persisting to disk and one for persisting to etcd)

File: tendrl/common/atoms/base_atom.py (hunk 1 above)
Comment (shtripat) [inline tendrl/common/atoms/base_atom.py]: Done

File: tendrl/common/flows/base_flow.py (hunk 5 above)
Comment (shtripat) [inline tendrl/common/flows/base_flow.py]: Done

File: tendrl/common/config.py (hunk 4 above)
Comment (shtripat) [inline tendrl/common/config.py]: Done

File: tendrl/common/flows/flow.py (hunk 6 above)
Comment (shtripat) [inline tendrl/common/flows/flow.py]: Done

File: tendrl/common/flows/flow_execution_exception.py (hunk 7 above)
Comment (shtripat) [inline tendrl/common/flows/flow_execution_exception.py]: Done

File: tendrl/common/manager/manager.py (hunk 8 above)
Comment (shtripat) [inline tendrl/common/manager/manager.py]: ok. would try to have these

File: tendrl/common/manager/rpc.py (hunk 9 above)
Comment (shtripat) [inline tendrl/common/manager/rpc.py]: Done

File: tendrl/common/persistence/persister.py (hunk 10 above)
Comment (shtripat) [inline tendrl/common/persistence/persister.py]: ok. will do this

File: tendrl/common/flows/flow.py (hunk 6 above)
Comment (shtripat) [inline tendrl/common/flows/flow.py]: If we remove flow.py and move everything in base_flow.py, this class already has got a run() function and there is no way we can impose all subclasses to implement run() function. Is there any other way to achieve this?

**Comment by shtripat** [comment]
This PR is meant for https://github.com/Tendrl/specifications/issues/31

**Comment by shtripat** [comment]
@r0h4n @brainfunked @nthomas-redhat Incorporated the review comments. Plz have look.

File: tendrl/common/flows/flow.py (hunk 6 above)
Comment (r0h4n) [inline tendrl/common/flows/flow.py]: It is doable, check https://pymotw.com/2/abc/#concrete-methods-in-abcs

File: tendrl/common/manager/job_sync.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+    def validate_flow(self, raw_job):
+        definitions = yaml.load(
+            self.client.read(
+                self.syncJobThread._manager.defs_dir
+            ).value.decode("utf-8")
+        )
+        definitions = DefinitionsSchemaValidator(
+            definitions).sanitize_definitions()
+        # resp, msg = JobValidator(definitions).validateApi(raw_job)
```
Comment (r0h4n) [inline tendrl/common/manager/job_sync.py]: These validations need to be uncommented and enable or completely removed

File: tendrl/common/manager/job_sync.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+                            job, self.config, definitions).run()
+
+    def extract_flow_details(self, flow_name, definitions):
+        namespace = flow_name.split(".flows.")[0]
+        flow = definitions[namespace]['flows'][flow_name.split(".")[-1]]
+        return flow['atoms'], flow['help'], flow['enabled'], flow['inputs'], \
+            flow['pre_run'], flow['post_run'], flow['type'], flow['uuid']
+
+
+class SyncJobThread(gevent.greenlet.Greenlet):
```
Comment (r0h4n) [inline tendrl/common/manager/job_sync.py]: call this RpcJobProcessThread

File: tendrl/common/manager/manager.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        config,
+        events,
+        persister,
+        defs_dir
+    ):
+        self.name = name
+        self._config = config
+        self.integration_id = integration_id
+        self._complete = gevent.event.Event()
+        self._user_request_thread = SyncJobThread(self)
```
Comment (r0h4n) [inline tendrl/common/manager/manager.py]: rename user_request_thread to something like rpc_job_process_thread

File: tendrl/common/manager/manager.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        self._discovery_thread = events
+        self.persister = persister
```
Comment (r0h4n) [inline tendrl/common/manager/manager.py]: call this persister_thread

File: tendrl/common/atoms/base_atom.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+        inputs,
+        outputs,
+        uuid
+    ):
+        self.name = name
+        self.enabled = enabled
+        self.inputs = inputs
+        self.help = help
+        self.outputs = outputs
+        uuid = uuid
```
Comment (shtripat) [inline tendrl/common/atoms/base_atom.py]: Ah just missed ;)

File: tendrl/common/manager/job_sync.py (hunk 11 above)
Comment (shtripat) [inline tendrl/common/manager/job_sync.py]: Will be enabled once validator is corrected and works fine. @TimothyAsir is working on the same.

File: tendrl/common/manager/manager.py (hunk 13 above)
Comment (shtripat) [inline tendrl/common/manager/manager.py]: Done

File: tendrl/common/manager/manager.py (hunk 14 above)
Comment (shtripat) [inline tendrl/common/manager/manager.py]: Done

File: tendrl/common/manager/rpc_job_process.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+    def validate_flow(self, raw_job):
+        definitions = yaml.load(
+            self.client.read(
+                self.syncJobThread._manager.defs_dir
+            ).value.decode("utf-8")
+        )
+        definitions = DefinitionsSchemaValidator(
+            definitions).sanitize_definitions()
+        # resp, msg = JobValidator(definitions).validateApi(raw_job)
```
Comment (r0h4n) [inline tendrl/common/manager/rpc_job_process.py]: Either remove the commented code or implement it

File: tendrl/common/tests/test_config.py [hunk 17]
```diff
@@ -8,11 +8,9 @@ def setup_method(self, method):
         config.os.path.exists = MagicMock(return_value=True)
 
     def test_Tendrlconfig(self, monkeypatch):
-        config.TendrlConfig()
+        config.TendrlConfig("module", "/etc/tendrl/tendrl.conf")
```
Comment (r0h4n) [inline tendrl/common/tests/test_config.py]: tests should not assume local files existence, please create a temp conf and supply that, dont hardcode

File: tendrl/common/tests/test_config.py [hunk 18]
```diff
@@ -8,11 +8,9 @@ def setup_method(self, method):
...
         config.os.path.exists.assert_called_with("/etc/tendrl/tendrl.conf")
         monkeypatch.setattr(config.os, 'environ', {"TENDRL_CONFIG": "/temp/"})
-        config.TendrlConfig()
-        config.os.path.exists.assert_called_with("/temp/")
         config.os.path.exists = MagicMock(return_value=False)
         with pytest.raises(config.ConfigNotFound):
-            config.TendrlConfig()
+            config.TendrlConfig("module", "/temp/dummy.conf")
```
Comment (r0h4n) [inline tendrl/common/tests/test_config.py]: when and who is supposed to create /temp/dummy.conf?

File: tendrl/common/manager/rpc_job_process.py (hunk 16 above)
Comment (shtripat) [inline tendrl/common/manager/rpc_job_process.py]: Will remove for the time being

File: tendrl/common/tests/test_config.py (hunk 17 above)
Comment (shtripat) [inline tendrl/common/tests/test_config.py]: Its not actually going to look for the file name passed here. test is just about making sure functions are called with passed arguments.

File: tendrl/common/tests/test_config.py (hunk 18 above)
Comment (shtripat) [inline tendrl/common/tests/test_config.py]: Same as above

**Comment by shtripat** [comment]
> Fix Travis failures

@r0h4n regarding travis failure, this is because of dynamic loading of module and atom using exec. Not sure how that could be fixed.

**Comment by r0h4n** [comment]
There are more than one travis failures, please fix them.

Also, figure out the "the_atome" failure issue via flake8/pep8 docs

**Comment by shtripat** [comment]
Verified the base gluster import flow with latest code and works fine.

File: tendrl/common/config.py [hunk 19]
```diff
@@ -9,22 +9,17 @@ class ConfigNotFound(Exception):
...
         ConfigParser.SafeConfigParser.__init__(self)
 
-        try:
-            self.path = os.environ[CONFIG_PATH_VAR]
-        except KeyError:
-            self.path = DEFAULT_CONFIG_PATH
+        self.path = cfg_file_path
 
         if not os.path.exists(self.path):
-            err = ConfigNotFound("Bridge_Common Configuration not found at "
```
Comment (shtripat) [inline tendrl/common/config.py]: done

File: tendrl/common/flows/base_flow.py [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+        LOG.info("Starting execution of pre-runs for flow: %s" %
+                 self.job['run'])
+        if self.pre_run is not None:
+            for mod in self.pre_run:
+                ret_val = self.execute_atom(mod)
+
+                if not ret_val:
+                    LOG.error("Failed executing pre-run: %s for flow: %s" %
+                              (pre_atom, self.job['run']))
+                    raise Exception(
```
Comment (shtripat) [inline tendrl/common/flows/base_flow.py]: Thats correct. We can use FlowExecutionFailedError. Will change that.

File: tendrl/common/flows/base_flow.py [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+
+    def extract_atom_details(self, atom_name):
+        namespace = atom_name.split('.objects.')[0]
+        object_name = atom_name.split('.objects.')[1].split('.atoms.')[0]
+        atoms = self.definitions[namespace]['objects'][object_name]['atoms']
+        atom = atoms[atom_name.split('.')[-1]]
+        return atom.get('name'), atom.get('enabled'), atom.get('help'), \
+            atom.get('inputs'), atom.get('outputs'), atom.get('uuid')
+
+    def execute_atom(self, mod):
```
Comment (shtripat) [inline tendrl/common/flows/base_flow.py]: `mod` is nothing but the full module name. Will add some comments to explain the same here.

File: tendrl/common/flows/base_flow.py [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+        class_name = utils.to_camel_case(mod.split(".")[-1])
+        if "tendrl" in mod and "atoms" in mod:
+            atom_name, enabled, help, inputs, outputs, uuid = \
+                self.extract_atom_details(mod)
+            exec("from %s import %s as the_atom" % (
+                mod.lower().strip("."),
+                class_name.strip("."))
+            )
+
+            ret_val = the_atom(    # noqa: F821
```
Comment (shtripat) [inline tendrl/common/flows/base_flow.py]: We may think of having a AtomExecutionFailedError defined and individual atoms can use them for any exception cases and in base_flow we can catch that. I can add one here..

File: tendrl/common/log.py [hunk 23]
```diff
@@ -20,9 +16,3 @@ def setup_logging(
         logging.config.dictConfig(log_config)
     else:
         raise Exception("logging configuration not found")
```
Comment (shtripat) [inline tendrl/common/log.py]: I understand but there is no generic exception defined for this scenario. We would need one for this. Can be created in this source only may be.

**Pull request merged**